lexicon/providers/aliyun.py

```diff
@@ -0,0 +1,217 @@
+
+        LOGGER.debug('update_record: %s', identifier)
+
+        query_params = {'RecordId': identifier}
+
+        if rtype:
+            query_params['Type'] = rtype
+
+        if name:
+            query_params['RR'] = self._relative_name(name)
+
+        if content:
+            query_params['Value'] = content
+
+        query_params['TTL'] = self._get_lexicon_option('ttl')
+        self._request('UpdateDomainRecord', query_params=query_params)
+
+        return True
+
+    # Delete an existing record.
+    # If record does not exist, do nothing.
+    def _delete_record(self, identifier=None, rtype=None, name=None, content=None):
+        delete_resource_id = []
+        if not identifier:
+            resources = self._list_records(rtype, name, content)
+            delete_resource_id = [resource['id'] for resource in resources]
+        else:
+            delete_resource_id.append(identifier)
+
+        LOGGER.debug('delete_records: %s', delete_resource_id)
+
+        for resource_id in delete_resource_id:
+            self._request('DeleteDomainRecord', query_params={'RecordId': resource_id})
+
+        return True
+
+    def _request(self, action='GET', url='/', data=None, query_params=None):
+        # force set to the fixed api endpoint
+        url = ALIYUN_DNS_API_ENDPOINT
+
+        if query_params is None:
+            query_params = {}
+
+        query_params.update(self._build_default_query_params(action))
+        query_params.update(self._build_signature_parameters())
+
+        query_params.update({'Signature': self._calculate_signature('GET', query_params)})
+
+        response = requests.request('GET', url, params=query_params)
+
+        # TODO: add exception handle
+        return response.json()
+
+    # @param query_params should be {} and not none
+    def _calculate_signature(self, http_method, query_params):
+        access_secret = self._get_provider_option('access_secret')
+        sign_secret = access_secret+'&'
+
+        query_list = list(query_params.items())
+        query_list.sort(key=lambda t: t[0])
+
+        canonicalized_query_string = urllib.parse.urlencode(query_list)
+
+        string_to_sign = '&'.join([
+            http_method,
+            urllib.parse.quote_plus('/'),
+            urllib.parse.quote_plus(canonicalized_query_string)])
+
+        signature = ""
+        if sys.version_info.major > 2:
+            import base64
+
+            sign_secret_bytes = bytes(sign_secret, 'utf-8')
+            string_to_sign_bytes = bytes(string_to_sign, 'utf-8')
+            sign = hmac.new(sign_secret_bytes, string_to_sign_bytes, sha1)
+            signature = base64.b64encode(sign.digest()).decode()
+        else:
+            sign = hmac.new(sign_secret, string_to_sign, sha1)
+            signature = sign.digest().encode('base64').rstrip('\n')
+
+        return signature
+
+    def _build_signature_parameters(self):
+        access_key_id = self._get_provider_option('access_key_id')
```

Comment:
Here also, `access_key_id` may be `None`, raising an exception in this case will avoid to build falsy signatures parameters.